src/client/shader.cpp

```diff
@@ -620,15 +653,62 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
 		return shaderinfo;
 
 	// Create shaders header
-	std::string shaders_header = "#version 120\n";
-
+	bool use_gles = false;
+#if ENABLE_GLES
+	use_gles = driver->getDriverType() == video::EDT_OGLES2;
+#endif
+	std::string shaders_header, vertex_header, pixel_header; // geometry shaders aren’t supported in GLES<3
+	if (use_gles) {
+		shaders_header =
+			"#version 100\n"
+			;
+		vertex_header = R"(
+			uniform highp mat4 mWorldView;
+			uniform highp mat4 mWorldViewProj;
+			uniform mediump mat4 mTexture;
+			uniform mediump mat3 mNormal;
+
+			attribute highp vec4 inVertexPosition;
+			attribute lowp vec4 inVertexColor;
+			attribute mediump vec4 inTexCoord0;
+			attribute mediump vec3 inVertexNormal;
+			attribute mediump vec4 inVertexTangent;
+			attribute mediump vec4 inVertexBinormal;
+			)";
+		pixel_header = R"(
+			precision mediump float;
+			)";
+	} else {
+		shaders_header = R"(
+			#version 120
+			#define lowp
+			#define mediump
+			#define highp
```

Comment:
https://www.khronos.org/opengl/wiki/Type_Qualifier_(GLSL)#Precision_qualifiers

According to this lowp/mediump/highp are valid but don't have any meaning, so no need to define them.

Reply:
They are valid in GLSL 1.4 (OpenGL 3) but MT uses GLSL 1.2 (OpenGL 2). They are reserved and not valid there.

Reply:
Argh, they are reserved... so I shouldn’t define them either, according to the spec. Things become hairy...

Reply:
Oops, didn't check the version. This is a problem indeed.

If that doesn't work out the code could fall back to text substitution (or compiling with `#version 140` if supported?).

Reply:
It works on my system ofc (and doesn’t work without the defines) though, so this compiler (mesa) is sane.